src/Workspaces/Remote/ServiceHub/Services/BrokeredServiceBase.cs

```diff
@@ -0,0 +1,209 @@
+               IServiceProvider hostProvidedServices,
+               ServiceActivationOptions serviceActivationOptions,
+               IServiceBroker serviceBroker,
+               AuthorizationServiceClient? authorizationServiceClient)
+            {
+                // Dispose the AuthorizationServiceClient since we won't be using it
+                authorizationServiceClient?.Dispose();
+
+                return Task.FromResult((object)Create(
+                    stream.UsePipe(),
+                    hostProvidedServices,
+                    serviceActivationOptions,
+                    serviceBroker));
+            }
+
+            object IFactory.Create(IDuplexPipe pipe, IServiceProvider hostProvidedServices, ServiceActivationOptions serviceActivationOptions, IServiceBroker serviceBroker)
+                => Create(pipe, hostProvidedServices, serviceActivationOptions, serviceBroker);
+
+            Type IFactory.ServiceType => typeof(TService);
+
+            internal TService Create(
+               IDuplexPipe pipe,
+               IServiceProvider hostProvidedServices,
+               ServiceActivationOptions serviceActivationOptions,
+               IServiceBroker serviceBroker)
+            {
+                var descriptor = ServiceDescriptors.GetServiceDescriptor(typeof(TService), isRemoteHost64Bit: IntPtr.Size == 8);
+                var serverConnection = descriptor.ConstructRpcConnection(pipe);
+
+                var args = new ServiceConstructionArguments(hostProvidedServices, serviceBroker, new CancellationTokenSource());
+                var service = CreateService(args, descriptor, serverConnection, serviceActivationOptions.ClientRpcTarget);
+
+                serverConnection.AddLocalRpcTarget(service);
+                serverConnection.StartListening();
+
+                return service;
+            }
+        }
+
+        internal abstract class FactoryBase<TService, TCallback> : FactoryBase<TService>
+            where TService : class
+            where TCallback : class
+        {
+            protected abstract TService CreateService(in ServiceConstructionArguments arguments, RemoteCallback<TCallback> callback);
+
+            protected sealed override TService CreateService(in ServiceConstructionArguments arguments)
+                => throw ExceptionUtilities.Unreachable;
+
+            protected sealed override TService CreateService(
+                in ServiceConstructionArguments arguments,
+                ServiceRpcDescriptor descriptor,
+                ServiceRpcDescriptor.RpcConnection serverConnection,
+                object? clientRpcTarget)
+            {
+                Contract.ThrowIfNull(descriptor.ClientInterface);
+                var callback = (TCallback)(clientRpcTarget ?? serverConnection.ConstructRpcClient(descriptor.ClientInterface));
+                return CreateService(arguments, new RemoteCallback<TCallback>(callback, arguments.ClientDisconnectedSource));
+            }
+        }
```

Comment:
would honeslty love these moved to their own files.

Reply:
Sounds good.